crates/spicepod/src/component/dataset.rs

```diff
@@ -366,3 +369,148 @@ pub mod replication {
         pub enabled: bool,
     }
 }
+
+pub mod column {
+    use std::collections::HashMap;
+
+    #[cfg(feature = "schemars")]
+    use schemars::JsonSchema;
+    use serde::{de::Error, Deserialize, Serialize};
+
+    use crate::component::embeddings::EmbeddingChunkConfig;
+
+    #[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+    #[cfg_attr(feature = "schemars", derive(JsonSchema))]
+    pub struct Column {
+        pub name: String,
+
+        /// Optional semantic details about the column
+        #[serde(default, skip_serializing_if = "Option::is_none")]
+        pub description: Option<String>,
+
+        #[serde(default, skip_serializing_if = "Vec::is_empty")]
+        pub embeddings: Vec<ColumnLevelEmbeddingConfig>,
+    }
+
+    impl Column {
+        /// Return the column-level metadata that should be added to a [`arrow::datatypes::Field`].
+        #[must_use]
+        pub fn metadata(&self) -> HashMap<String, String> {
+            let mut metadata = HashMap::new();
+            if let Some(d) = self.description.as_ref() {
+                metadata.insert("description".to_string(), d.to_string());
+            }
+            metadata
+        }
+    }
+
+    /// Configuration for if and how a dataset's column should be embedded.
+    /// Different to [`crate::component::embeddings::ColumnEmbeddingConfig`],
+    /// as [`ColumnLevelEmbeddingConfig`] should be a property of [`Column`],
+    /// not [`super::Dataset`].
+    ///
+    /// [`crate::component::embeddings::ColumnEmbeddingConfig`] will be
+    /// deprecated long term in favour of [`ColumnLevelEmbeddingConfig`].
+    ///
+    #[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+    #[cfg_attr(feature = "schemars", derive(JsonSchema))]
+    pub struct ColumnLevelEmbeddingConfig {
+        #[serde(rename = "from", default)]
+        pub model: String,
+
+        #[serde(skip_serializing_if = "Option::is_none")]
+        pub chunking: Option<EmbeddingChunkConfig>,
+
+        #[serde(
+            rename = "row_id",
+            default,
+            deserialize_with = "deserialize_row_ids",
+            skip_serializing_if = "Option::is_none"
+        )]
+        pub row_ids: Option<Vec<String>>,
+    }
+
+    // Let `row_id` handle single string or arrays. All acceptable
+    // ```yaml
+    // row_id: foo
+    //
+    // row_id: foo, bar
+    //
+    // row_id:
+    //  - foo
+    //  - bar
+    // ```
+    fn deserialize_row_ids<'de, D>(deserializer: D) -> Result<Option<Vec<String>>, D::Error>
```

Comment:
One of the issues with custom deserialization functions is that the errors they produce are sometimes not helpful. What errors occur here if the format isn't valid? Maybe that could be part of the tests below.

Reply:
Good idea.